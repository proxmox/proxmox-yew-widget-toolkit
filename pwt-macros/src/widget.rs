use proc_macro::TokenStream;
use quote::{format_ident, quote};
use syn::ext::IdentExt;
use syn::parse::{Parse, ParseStream};
use syn::spanned::Spanned;
use syn::{parse_macro_input, Data, DeriveInput};
use syn::{Attribute, Error, Fields, Ident, Path, Result, Token};

#[derive(Debug)]
pub(crate) struct WidgetSetup {
    pwt_crate_name: Option<Ident>,
    component_name: Option<Path>,
    is_input: bool,
    is_container: bool,
    is_element: bool,
    is_svg: bool,
}

impl Parse for WidgetSetup {
    fn parse(input: ParseStream) -> Result<Self> {
        let mut pwt_crate_name = None;
        let mut component_name = None;
        let mut is_input = false;
        let mut is_container = false;
        let mut is_element = false;
        let mut is_svg = false;

        loop {
            if input.is_empty() {
                break;
            }
            let lookahead = input.lookahead1();
            if lookahead.peek(Token![@]) {
                let _: Token![@] = input.parse()?;
                let mixin: Ident = input.parse()?;
                if mixin == "input" {
                    is_input = true;
                } else if mixin == "container" {
                    is_container = true;
                } else if mixin == "element" {
                    is_element = true;
                } else if mixin == "svg" {
                    is_svg = true;
                } else {
                    return Err(Error::new(mixin.span(), "no such widget mixin"));
                }

                if input.is_empty() {
                    break;
                }
            } else if lookahead.peek(Token![!]) {
                pwt_crate_name = Some(input.parse()?);
            } else {
                let opt: Ident = input.parse()?;
                let _: Token![=] = input.parse()?;
                match opt.to_string().as_ref() {
                    "pwt" => {
                        let name: Ident = input.call(Ident::parse_any)?;
                        if pwt_crate_name.is_some() {
                            return Err(Error::new(name.span(), "multiple pwt name definitions"));
                        }
                        pwt_crate_name = Some(name);
                    }
                    "comp" => {
                        let path: Path = input.parse()?;
                        if component_name.is_some() {
                            return Err(Error::new(path.span(), "multiple component definitions"));
                        }
                        component_name = Some(path);
                    }
                    _ => {
                        return Err(Error::new(opt.span(), "unknown widget option"));
                    }
                }
            }
            if input.is_empty() {
                break;
            }
            let _: Token![,] = input.parse()?;
            if input.is_empty() {
                break;
            }
        }

        Ok(WidgetSetup {
            pwt_crate_name,
            component_name,
            is_input,
            is_container,
            is_element,
            is_svg,
        })
    }
}

pub(crate) fn handle_widget_struct(setup: &WidgetSetup, input: TokenStream) -> TokenStream {
    let widget = parse_macro_input!(input as DeriveInput);

    derive_widget(setup, widget)
        .unwrap_or_else(syn::Error::into_compile_error)
        .into()
}

fn has_property_derive(attrs: &Vec<Attribute>) -> Result<bool> {
    // SIGH!!
    for attr in attrs {
        if attr.style != syn::AttrStyle::Outer {
            continue;
        }
        if let Some(ident) = attr.path().get_ident() {
            if ident != "derive" {
                continue;
            }
            let mut has_properties = false;
            attr.parse_nested_meta(|nested| {
                if nested.path.is_ident("Properties") {
                    has_properties = true;
                }
                Ok(())
            })?;

            return Ok(has_properties);
        }
    }

    Ok(false)
}

fn derive_widget(setup: &WidgetSetup, widget: DeriveInput) -> Result<proc_macro2::TokenStream> {
    let DeriveInput {
        attrs,
        vis,
        ident,
        generics,
        data,
    } = widget;

    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();

    let has_property_derive = has_property_derive(&attrs)?;

    let pwt: Ident = setup.pwt_crate_name.clone().unwrap_or(format_ident!("pwt"));

    let fields = match data {
        Data::Struct(data) => match data.fields {
            Fields::Named(fields) => fields,
            _ => {
                return Err(Error::new(
                    data.struct_token.span,
                    "expected `struct` with named fields",
                ));
            }
        },
        Data::Enum(data) => {
            return Err(Error::new(data.enum_token.span, "expected `struct`"));
        }
        Data::Union(data) => {
            return Err(Error::new(data.union_token.span, "expected `struct`"));
        }
    };

    let fields = fields.named; // remove braces

    let mut opt_fields: Vec<_> = Vec::new();

    let prop_or_default = if has_property_derive {
        quote! { #[prop_or_default] }
    } else {
        quote! {}
    };

    if setup.is_input {
        opt_fields.push(quote! {
            #[doc(hidden)]
            #prop_or_default
            pub input_props: #pwt::props::FieldStdProps,
        });
    }

    if setup.is_container {
        opt_fields.push(quote! {
            #[doc(hidden)]
            #prop_or_default
            pub children: Vec<::yew::virtual_dom::VNode>,
        });
    }

    if setup.is_element {
        opt_fields.push(quote! {
            #[doc(hidden)]
            #prop_or_default
            pub listeners: #pwt::props::ListenersWrapper,
        });
    }

    let mut output = quote! {
        #(#attrs)*
        #vis struct #ident #generics {

            #[doc(hidden)]
            #prop_or_default
            pub std_props: #pwt::props::WidgetStdProps,

            #(#opt_fields)*

            #fields
        }
    };

    output.extend(quote! {
        impl #impl_generics #pwt::props::WidgetBuilder for #ident #ty_generics #where_clause {
            fn as_std_props_mut(&mut self) -> &mut #pwt::props::WidgetStdProps {
                &mut self.std_props
            }
        }
    });

    if !setup.is_svg {
        output.extend(quote!{
            impl #impl_generics #pwt::props::AsClassesMut for #ident #ty_generics #where_clause {
                fn as_classes_mut(&mut self) -> &mut ::yew::Classes {
                    &mut self.std_props.class
                }
            }
            impl #impl_generics #pwt::props::CssMarginBuilder for #ident #ty_generics #where_clause {}
            impl #impl_generics #pwt::props::CssPaddingBuilder for #ident #ty_generics #where_clause {}
            impl #impl_generics #pwt::props::CssBorderBuilder for #ident #ty_generics #where_clause {}
        });
    }

    if setup.is_element {
        output.extend(quote! {
            impl #impl_generics #pwt::props::EventSubscriber for #ident #ty_generics #where_clause {
                fn as_listeners_mut(&mut self) -> &mut #pwt::props::ListenersWrapper {
                    &mut self.listeners
                }
            }
        });
    }

    if setup.is_container {
        output.extend(quote!{
            impl #impl_generics #pwt::props::ContainerBuilder for #ident #ty_generics #where_clause {
                fn as_children_mut(&mut self) -> &mut Vec<::yew::virtual_dom::VNode> {
                    &mut self.children
                }
            }
        });
    }

    if setup.is_input {
        output.extend(quote! {
            impl #impl_generics #pwt::props::FieldBuilder for #ident #ty_generics #where_clause {
                fn as_input_props_mut(&mut self) -> &mut #pwt::props::FieldStdProps {
                    &mut self.input_props
                }
                fn as_input_props(&self) -> & #pwt::props::FieldStdProps {
                    &self.input_props
                }
            }
        });
    }

    if let Some(component_name) = &setup.component_name {
        output.extend(quote!{
            impl #impl_generics Into<::yew::virtual_dom::VNode> for #ident #ty_generics #where_clause {
                fn into(self) -> ::yew::virtual_dom::VNode {
                    let key = self.std_props.key.clone();
                    let comp = ::yew::virtual_dom::VComp::new::<#component_name>(
                        ::std::rc::Rc::new(self), key,
                    );
                    ::yew::virtual_dom::VNode::from(comp)
                }
            }
        });
    } else {
        output.extend(quote!{
            impl #impl_generics Into<::yew::virtual_dom::VNode> for #ident #ty_generics #where_clause {
                fn into(self) -> ::yew::virtual_dom::VNode {
                    let vtag: ::yew::virtual_dom::VTag = self.into();
                    ::yew::virtual_dom::VNode::from(vtag)
                }
            }
        });
    }

    //eprintln!("TEST {}", output);
    Ok(output)
}
