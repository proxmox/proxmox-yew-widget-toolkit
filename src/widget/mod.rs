//! # Basic widgets

mod action_icon;
pub use action_icon::{ActionIcon, PwtActionIcon};

mod alert_dialog;
pub use alert_dialog::{AlertDialog, PwtAlertDialog};

mod message_box;
pub use message_box::{MessageBox, MessageBoxButtons, PwtMessageBox};

pub mod align;

mod button;
pub use button::{Button, PwtButton};

pub mod canvas;

mod card;
pub use card::Card;

mod catalog_loader;
pub use catalog_loader::{CatalogLoader, PwtCatalogLoader};

mod column;
pub use column::Column;

mod container;
pub use container::Container;

mod desktop_app;
pub use desktop_app::{DesktopApp, PwtDesktopApp};

mod dropdown;
pub use dropdown::{Dropdown, PwtDropdown, RenderDropdownPickerFn};

pub mod dom;

mod fa;
pub use fa::{Fa, PwtFa};

pub mod focus;

pub mod form;

mod input;
pub use input::Input;

mod input_panel;
pub use input_panel::InputPanel;

mod language_selector;
pub use language_selector::{LanguageSelector, ProxmoxLanguageSelector};

mod mask;
pub use mask::{Mask, PwtMask};

mod mini_scroll;
pub use mini_scroll::{MiniScroll, MiniScrollMode, PwtMiniScroll};

pub mod nav;

pub mod menu;

mod meter;
pub use meter::Meter;

mod split_pane;
pub use split_pane::{Pane, PwtSplitPane, SplitPane};

pub mod data_table;

mod dialog;
pub use dialog::{Dialog, PwtDialog};

mod panel;
pub use panel::Panel;

mod grid_picker;
pub use grid_picker::GridPicker;

mod progress;
pub use progress::Progress;

mod row;
pub use row::Row;

mod rtl_switcher;
pub use rtl_switcher::RtlSwitcher;

mod selection_view;
pub use selection_view::{PwtSelectionView, SelectionView, SelectionViewRenderInfo};

mod segmented_button;
pub use segmented_button::{PwtSegmentedButton, SegmentedButton};

mod size_observer;
pub use size_observer::{IntoSizeCallback, SizeObserver};

mod tab;
pub use tab::*;

mod theme_loader;
pub use theme_loader::ThemeLoader;

mod theme_density_selector;
pub use theme_density_selector::{PwtThemeDensitySelector, ThemeDensitySelector};

mod theme_mode_selector;
pub use theme_mode_selector::ThemeModeSelector;

mod theme_name_selector;
pub use theme_name_selector::ThemeNameSelector;

mod toolbar;
pub use toolbar::{PwtToolbar, Toolbar};

mod tooltip;
pub use tooltip::{PwtTooltip, Tooltip};

//mod virtual_scroll;
//pub use virtual_scroll::VirtualScroll;

mod visibility_observer;
pub use visibility_observer::VisibilityObserver;

use std::sync::atomic::{AtomicUsize, Ordering};

static UNIQUE_ELEMENT_ID: AtomicUsize = AtomicUsize::new(0);

/// Creates an unique element ID (`PwtElementId{unique_num}`)
pub fn get_unique_element_id() -> String {
    let id = UNIQUE_ELEMENT_ID.fetch_add(1, Ordering::SeqCst);
    format!("PwtElementId{}", id)
}

/// Creates a nicely formatted error message.
pub fn error_message(text: &str, class: &str) -> yew::Html {
    message_box::message(text, class, "fa-exclamation-triangle")
}
