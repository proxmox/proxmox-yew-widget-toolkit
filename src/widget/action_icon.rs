use std::rc::Rc;

use yew::prelude::*;
use yew::virtual_dom::{VComp, VNode, Key};
use yew::html::{IntoEventCallback, IntoPropValue};

use crate::prelude::IntoOptionalKey;
use crate::props::{EventSubscriber, WidgetBuilder};
use crate::widget::Container;

/// A clickable icon. Like [Button](super::Button) without any decoration (inline element).
///
/// This component is useful in data tables because it is visually lighter than a button.
#[derive(Properties, PartialEq, Clone)]
pub struct ActionIcon {
    /// Yew component `ref`.
    #[prop_or_default]
    pub node_ref: NodeRef,

    /// Yew `key` property
    pub key: Option<Key>,

    /// CSS class
    #[prop_or_default]
    pub class: Classes,

    /// The CSS icon class
    pub icon_class: Option<Classes>,

    /// Html tabindex (defaults to -1)
    pub tabindex: Option<i32>,

    /// Aria label
    pub aria_label: Option<AttrValue>,

    /// Disable flag
    #[prop_or_default]
    pub disabled: bool,

    /// Activate callback (click, enter, space)
    pub on_activate: Option<Callback<()>>,
}

impl ActionIcon {

    /// Create a new instance.
    pub fn new(icon_class: impl Into<Classes>) -> Self {
        yew::props!(Self { icon_class: icon_class.into() })
    }

    /// Builder style method to set the yew `node_ref`
    pub fn node_ref(mut self, node_ref: NodeRef) -> Self {
        self.set_node_ref(node_ref);
        self
    }

    /// Method to set the yew `node_ref`
    pub fn set_node_ref(&mut self, node_ref: NodeRef) {
        self.node_ref = node_ref;
    }

    /// Builder style method to set the yew `key` property
    pub fn key(mut self, key: impl IntoOptionalKey) -> Self {
        self.set_key(key);
        self
    }

    /// Method to set the yew `key` property
    pub fn set_key(&mut self, key: impl IntoOptionalKey) {
        self.key = key.into_optional_key();
    }

    /// Builder style method to add a html class
    pub fn class(mut self, class: impl Into<Classes>) -> Self {
        self.add_class(class);
        self
    }

    /// Method to add a html class
    pub fn add_class(&mut self, class: impl Into<Classes>) {
        self.class.push(class);
    }

    /// Builder style method to set the html tabindex attribute
    pub fn tabindex(mut self, index: impl IntoPropValue<Option<i32>>) -> Self {
        self.set_tabindex(index);
        self
    }

    /// Method to set the html tabindex attribute
    pub fn set_tabindex(&mut self, index: impl IntoPropValue<Option<i32>>) {
        self.tabindex = index.into_prop_value();
    }

    /// Builder style method to set the html aria-label attribute
    pub fn aria_label(mut self, label: impl IntoPropValue<Option<AttrValue>>) -> Self {
        self.set_aria_label(label);
        self
    }

    /// Method to set the html aria-label attribute
    pub fn set_aria_label(&mut self, label: impl IntoPropValue<Option<AttrValue>>) {
        self.aria_label = label.into_prop_value();
    }

    /// Builder style method to set the disabled flag
    pub fn disabled(mut self, disabled: bool) -> Self {
        self.set_disabled(disabled);
        self
    }

    /// Method to set the disabled flag
    pub fn set_disabled(&mut self, disabled: bool) {
        self.disabled = disabled;
    }

    /// Builder style method to set the activate callback.
    pub fn on_activate(mut self, cb: impl IntoEventCallback<()>) -> Self {
        self.on_activate = cb.into_event_callback();
        self
    }
}

#[doc(hidden)]
pub struct PwtActionIcon;

impl Component for PwtActionIcon {
    type Message = ();
    type Properties = ActionIcon;

    fn create(_ctx: &Context<Self>) -> Self {
        Self {}
    }

    fn view(&self, ctx: &Context<Self>) -> Html {
        let props = ctx.props();

        let disabled = props.disabled;

        let tabindex = match props.tabindex {
            Some(tabindex) => format!("{tabindex}"),
            None => String::from("-1"),
        };

        Container::new()
            .tag("i")
            .node_ref(props.node_ref.clone())
            .attribute("tabindex", (!disabled).then(|| tabindex))
            .attribute("role", "button")
            .attribute("aria-label", props.aria_label.clone())
            .class("pwt-action-icon")
            .class(props.disabled.then(|| "disabled"))
            .class(props.class.clone())
            .class(props.icon_class.clone())
            .onclick({
                let on_activate = props.on_activate.clone();
                move |event: MouseEvent| {
                    event.stop_propagation();
                    if disabled { return; }
                    if let Some(on_activate) = &on_activate {
                        on_activate.emit(());
                    }
                }
            })
            .onkeydown({
                let on_activate = props.on_activate.clone();
                move |event: KeyboardEvent| {
                    match event.key().as_ref() {
                        "Enter" | " " => {
                            event.stop_propagation();
                            if disabled { return; }
                            if let Some(on_activate) = &on_activate {
                                on_activate.emit(());
                            }
                        }
                        _ => {}
                    }
                }
            })
            // suppress double click to avoid confusion when used inside tables/trees
            .ondblclick(move |event: MouseEvent| { event.stop_propagation(); })
            .into()
    }
}

impl Into<VNode> for ActionIcon {
    fn into(self) -> VNode {
        let key = self.key.clone();
        let comp = VComp::new::<PwtActionIcon>(Rc::new(self), key);
        VNode::from(comp)
    }
}
