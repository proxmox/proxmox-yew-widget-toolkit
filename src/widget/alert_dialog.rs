use std::rc::Rc;

use yew::html::{IntoEventCallback, IntoPropValue};
use yew::prelude::*;
use yew::virtual_dom::{VComp, VNode};

use crate::prelude::*;
use crate::widget::{Button, Dialog, Row, Toolbar};

/// Alert Dialog - Modal window to display error messages.
#[derive(Clone, Properties, PartialEq)]
pub struct AlertDialog {
    /// Optional dialog title - defaults to "Alert".
    pub title: Option<AttrValue>,
    /// The error message.
    pub message: String,
    /// Close window callback.
    pub on_close: Option<Callback<()>>,
    /// Enable/disable dragging
    #[prop_or_default]
    pub draggable: bool,
}

impl AlertDialog {
    /// Create a new instance.
    pub fn new(message: impl Into<String>) -> Self {
        yew::props!(AlertDialog {
            message: message.into()
        })
    }

    /// Builder style method to set the dialog title.
    pub fn title(mut self, title: impl IntoPropValue<Option<AttrValue>>) -> Self {
        self.set_title(title);
        self
    }

    /// Method to set the dialog title.
    pub fn set_title(&mut self, title: impl IntoPropValue<Option<AttrValue>>) {
        self.title = title.into_prop_value();
    }

    /// Builder style method to set the window close callback.
    pub fn on_close(mut self, cb: impl IntoEventCallback<()>) -> Self {
        self.on_close = cb.into_event_callback();
        self
    }

    /// Builder style method to enable/disable dragging
    pub fn draggable(mut self, draggable: bool) -> Self {
        self.set_draggable(draggable);
        self
    }

    /// Enable/disable dragging
    pub fn set_draggable(&mut self, draggable: bool) {
        self.draggable = draggable;
    }
}

/// Creates a nicely formated error message.
pub fn error_message(text: &str, class: &str) -> Html {
    let icon_class = "pwt-bg-color-surface pwt-color-error pwt-p-2 pwt-shape-circle \
                      fa fa-lg fa-align-center fa-exclamation-triangle pwt-me-2";

    Row::new()
        .padding(2)
        .class(class.to_owned())
        .class("pwt-align-items-center")
        .attribute("style", "max-width:600px;")
        .with_child(html! {<span class={icon_class} aria-hidden="true"/>})
        .with_child(html! {<p>{text}</p>})
        .into()
}

/* we currently do not need this */
/*
pub fn display_load_result2<T>(result: &Option<Result<T, Error>>, render: impl Fn(&T) -> Html) -> Html {
    match result {
        None => html!{
            <div class="pwt-text-center pwt-p-4">
            {Fa::new("spinner").class("pwt-me-1").pulse()}
            {"Loading..."}
            </div>
        },
        Some(Ok(data)) => render(data),
        Some(Err(err)) => error_message(&format!("Error: {}", err), "pwt-p-2"),
    }
}
*/

#[function_component(PwtAlertDialog)]
#[doc(hidden)]
pub fn pwt_alert_dialog(props: &AlertDialog) -> Html {
    let onclick = Callback::from({
        let on_close = props.on_close.clone();
        move |_| {
            if let Some(on_close) = &on_close {
                on_close.emit(());
            }
        }
    });

    let title = format!("{}", props.title.as_deref().unwrap_or("Alert"));

    Dialog::new(title.clone())
        .draggable(props.draggable)
        .on_close(props.on_close.clone())
        .with_child(error_message(&props.message, "pwt-p-4"))
        .with_child(
            Toolbar::new()
                .class("emphased pwt-border-top")
                .with_flex_spacer()
                .with_child(Button::new("Continue").onclick(onclick).autofocus(true)),
        )
        .into()
}

impl Into<VNode> for AlertDialog {
    fn into(self) -> VNode {
        let comp = VComp::new::<PwtAlertDialog>(Rc::new(self), None);
        VNode::from(comp)
    }
}
