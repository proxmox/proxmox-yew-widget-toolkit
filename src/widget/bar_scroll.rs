use yew::prelude::*;

use gloo_timers::callback::Timeout;

use crate::props::{ContainerBuilder, EventSubscriber, WidgetBuilder};
use crate::widget::{Container, SizeObserver};

use pwt_macros::widget;

/// Scroll content horizontally.
#[widget(pwt=crate, comp=PwtBarScroll, @element)]
#[derive(Clone, PartialEq, Properties)]
pub struct BarScroll {
    content: Html,
}

impl BarScroll {
    /// Creates a new instance.
    pub fn new(content: impl Into<Html>) -> Self {
        yew::props!(Self { content: content.into() })
    }
}

#[doc(hidden)]
pub enum Msg {
    ScrollResize(f64, f64),
    ContentResize(f64, f64),
    HandleResize(f64, f64),
    Scroll,
    ScrollStop,
    ScrollLeft,
    ScrollRight,
}

enum ScrollMode {
    None,
    Left,
    Right,
}

#[doc(hidden)]
pub struct PwtBarScroll {
    handle_ref: NodeRef,
    handle_size_observer: Option<SizeObserver>,
    scroll_ref: NodeRef,
    content_ref: NodeRef,
    content_size_observer: Option<SizeObserver>,
    scroll_size_observer: Option<SizeObserver>,
    width: f64,
    handle_width: f64,
    content_width: f64,
    content_height: f64,
    pos: f64,
    scroll_mode: ScrollMode,
    scroll_timeout: Option<Timeout>,
}

impl PwtBarScroll {

    fn set_scroll_timeout(&mut self, ctx: &Context<Self>) {
        if self.scroll_timeout.is_some() { return; }
        let link = ctx.link().clone();
        self.scroll_timeout = Some(Timeout::new(1, move || {
            link.send_message(Msg::Scroll);
        }));
    }
}

impl Component for PwtBarScroll {
    type Message = Msg;
    type Properties = BarScroll;

    fn create(_ctx: &Context<Self>) -> Self {
        Self {
            handle_ref: NodeRef::default(),
            scroll_ref: NodeRef::default(),
            content_ref: NodeRef::default(),
            handle_size_observer: None,
            content_size_observer: None,
            scroll_size_observer: None,
            width: 0f64,
            handle_width: 0f64,
            content_width: 0f64,
            content_height: 0f64,
            pos: 0f64,
            scroll_mode: ScrollMode::None,
            scroll_timeout: None,
        }
    }

    fn update(&mut self, ctx: &Context<Self>, msg: Self::Message) -> bool {
        match msg {
            Msg::ScrollLeft => {
                self.scroll_mode = ScrollMode::Left;
                self.set_scroll_timeout(ctx);
                false
            }
            Msg::ScrollRight => {
                self.scroll_mode = ScrollMode::Right;
                self.set_scroll_timeout(ctx);
                false
            }
            Msg::ScrollResize(width, _height) => {
                self.width = width;
                true
            }
            Msg::ContentResize(width, height) => {
                self.content_width = width;
                self.content_height = height;
                true
            }
            Msg::HandleResize(width, _height) => {
                self.handle_width = width;
                true
            }
            Msg::Scroll => {
                self.scroll_timeout = None;

                let el = match self.scroll_ref.cast::<web_sys::Element>() {
                    None => return false,
                    Some(el) => el,
                };

                let width = el.get_bounding_client_rect().width();
                let diff = self.content_width - width;

                if diff <= 0f64 {
                    el.set_scroll_left(0);
                    self.pos = 0.0;
                    self.scroll_timeout = None;
                    return true;
                }

                let inc = (1.0/diff)*2.0;

                match self.scroll_mode  {
                    ScrollMode::None => {
                        return true;
                    }
                    ScrollMode::Right => {
                        self.pos = (self.pos + inc).min(1.0);
                        el.set_scroll_left((diff*self.pos) as i32);
                        if self.pos == 1f64 {
                            self.scroll_mode = ScrollMode::None;
                            return true;
                        }
                    }
                    ScrollMode::Left => {
                        self.pos = (self.pos - inc).max(0f64);
                        el.set_scroll_left((diff*self.pos) as i32);
                        if self.pos == 0f64 {
                            self.scroll_mode = ScrollMode::None;
                            return true;
                        }
                    }
                }

                self.set_scroll_timeout(ctx);

                true
            }
            Msg::ScrollStop => {
                self.scroll_mode = ScrollMode::None;
                self.scroll_timeout = None;
                true
            }
        }
    }

    fn view(&self, ctx: &Context<Self>) -> Html {
        let props = ctx.props();

        let content = Container::new()
            .node_ref(self.content_ref.clone())
            .class("pwt-position-absolute")
            .with_child(props.content.clone());

       let scroll = Container::new()
            .node_ref(self.scroll_ref.clone())
            .attribute("style", format!(
                "position:relative;cursor:{};overflow:hidden;flex:1 1 auto;height:{}px;",
                if self.scroll_timeout.is_some() { "wait" } else { "normal" },
                self.content_height,
            ))
            .with_child(content);

        let arrow_visible = (self.width + 2.0*self.handle_width) < self.content_width;

        let left = Container::new()
            .node_ref(self.handle_ref.clone())
            .class("pwt-autoscroll-left-arrow")
            .class(arrow_visible.then(|| "visible"))
            .class((self.pos <= 0.0).then(|| "disabled"))
            .with_child(html!{<i class="fa fa-chevron-left"/>})
            .onmousedown(ctx.link().callback(|_| Msg::ScrollLeft))
            .onmouseout(ctx.link().callback(|_| Msg::ScrollStop))
            .onmouseup(ctx.link().callback(|_| Msg::ScrollStop));

        let right = Container::new()
            .class("pwt-autoscroll-right-arrow")
            .class(arrow_visible.then(|| "visible"))
            .class((self.pos >= 1.0).then(|| "disabled"))
            .with_child(html!{<i class="fa fa-chevron-right"/>})
            .onmousedown(ctx.link().callback(|_| Msg::ScrollRight))
            .onmouseout(ctx.link().callback(|_| Msg::ScrollStop))
            .onmouseup(ctx.link().callback(|_| Msg::ScrollStop));

        yew::props!(Container {
            std_props: props.std_props.clone(),
            listeners: props.listeners.clone(),
        })
            .class("pwt-autoscroll")
            .with_child(left)
            .with_child(scroll)
            .with_child(right)
            .into()

    }

    fn rendered(&mut self, ctx: &Context<Self>, first_render: bool) {
        if first_render {
            if let Some(el) = self.scroll_ref.cast::<web_sys::Element>() {
                let link = ctx.link().clone();
                let size_observer = SizeObserver::new(&el, move |(width, height)| {
                    link.send_message(Msg::ScrollResize(width, height));
                });
                self.scroll_size_observer = Some(size_observer);
            }
            if let Some(el) = self.content_ref.cast::<web_sys::Element>() {
                let link = ctx.link().clone();
                let size_observer = SizeObserver::new(&el, move |(width, height)| {
                    link.send_message(Msg::ContentResize(width, height));
                });
                self.content_size_observer = Some(size_observer);
            }
            if let Some(el) = self.handle_ref.cast::<web_sys::Element>() {
                let link = ctx.link().clone();
                let size_observer = SizeObserver::new(&el, move |(width, height)| {
                    link.send_message(Msg::HandleResize(width, height));
                });
                self.handle_size_observer = Some(size_observer);
            }
        }

    }
}
