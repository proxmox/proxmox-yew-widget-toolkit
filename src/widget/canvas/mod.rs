mod circle;
pub use circle::Circle;

mod group;
pub use group:: Group;

mod hyperlink;
pub use hyperlink::Hyperlink;

mod line;
pub use line::Line;

mod polygon;
pub use polygon::Polygon;

mod path;
pub use path::Path;

mod polyline;
pub use polyline::Polyline;

mod rect;
pub use rect::Rect;

mod text;
pub use text::Text;

mod tspan;
pub use tspan::TSpan;

use std::fmt::Display;
use std::borrow::Cow;

use yew::prelude::*;
use yew::html::IntoPropValue;
use yew::virtual_dom::VTag;

use crate::props::WidgetBuilder;

use pwt_macros::widget;

#[derive(Copy, Clone, PartialEq)]
pub enum SvgLength {
    Px(f32),
    Em(f32),
    Fraction(f32),
}

impl Default for SvgLength {
    fn default() -> Self {
        SvgLength::Px(0.0)
    }
}

impl Display for SvgLength {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
        match self {
            SvgLength::Px(v) => write!(f, "{v}px"),
            SvgLength::Em(v) => write!(f, "{v}em"),
            SvgLength::Fraction(v) => write!(f, "{}%", v*100.0),
        }
    }
}

impl From<f32> for SvgLength {
    fn from(v: f32) -> SvgLength {
        SvgLength::Px(v)
    }
}

impl From<usize> for SvgLength {
    fn from(v: usize) -> SvgLength {
        SvgLength::Px(v as f32)
    }
}

impl From<i32> for SvgLength {
    fn from(v: i32) -> SvgLength {
        SvgLength::Px(v as f32)
    }
}

impl Into<AttrValue> for SvgLength {
    fn into(self) -> AttrValue {
        self.to_string().into()
    }
}

impl IntoPropValue<Option<AttrValue>> for SvgLength {
    fn into_prop_value(self) -> Option<AttrValue> {
        Some(self.into())
    }
}

/// SVG canvas (Html `<svg>` container).
#[widget(pwt=crate, @element, @container)]
#[derive(Properties, Clone, PartialEq)]
pub struct Canvas {}

impl Canvas {
    pub fn new() -> Self {
        yew::props!(Self {})
    }

    pub fn width(mut self, w: impl Into<SvgLength>) -> Self {
        self.set_width(w);
        self
    }

    pub fn set_width(&mut self, w: impl Into<SvgLength>) {
        self.set_attribute("width", w.into().to_string());
    }

    pub fn height(mut self, h: impl Into<SvgLength>) -> Self {
        self.set_height(h);
        self
    }

    pub fn set_height(&mut self, h: impl Into<SvgLength>) {
        self.set_attribute("height", h.into().to_string());
    }
}

impl Into<VTag> for Canvas {
    fn into(self) -> VTag {
        self.std_props.into_vtag(Cow::Borrowed("svg"), Some(self.listeners), Some(self.children))
    }
}
