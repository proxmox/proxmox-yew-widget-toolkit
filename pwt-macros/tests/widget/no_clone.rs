mod props {
    use yew::{virtual_dom::Key, Classes};

    pub trait FieldBuilder {
        fn as_input_props(&self) -> &FieldStdProps;
        fn as_input_props_mut(&mut self) -> &mut FieldStdProps;
    }
    pub trait CssBorderBuilder {}
    pub trait EventSubscriber: Sized {
        fn as_listeners_mut(&mut self) -> &mut ListenersWrapper;
    }
    pub trait CssPaddingBuilder {}
    pub trait CssMarginBuilder {}
    pub trait AsCssStylesMut {
        fn as_css_styles_mut(&mut self) -> &mut CssStyles;
    }

    /// Holds the CSS styles to set on elements
    #[derive(Clone, Default, Debug, PartialEq)]
    pub struct CssStyles {}

    pub trait WidgetStyleBuilder {}

    pub trait WidgetBuilder: Sized {
        fn as_std_props_mut(&mut self) -> &mut WidgetStdProps;
        fn as_std_props(&self) -> &WidgetStdProps;
    }

    pub trait AsClassesMut {
        fn as_classes_mut(&mut self) -> &mut Classes;
    }
    #[derive(PartialEq, Default, Clone)]
    pub struct FieldStdProps {}
    #[derive(PartialEq, Default, Clone)]
    pub struct WidgetStdProps {
        pub key: Option<Key>,
        pub class: Classes,
        pub styles: CssStyles,
    }
    #[derive(PartialEq, Default, Clone)]
    pub struct ListenersWrapper {}
}

use pwt_macros::widget;
use yew::prelude::*;

#[widget(pwt=crate, comp=FooComp, @input, @element)]
#[derive(Properties, PartialEq)]
struct Foo {}

impl Foo {
    fn new() -> Self {
        yew::props!(Self {})
    }
}

struct FooComp {}
impl Component for FooComp {
    type Message = ();
    type Properties = Foo;

    fn create(_ctx: &yew::Context<Self>) -> Self {
        Self {}
    }

    fn view(&self, _ctx: &yew::Context<Self>) -> yew::Html {
        html! {<div>{"Foo"}</div>}
    }
}

fn main() {}
