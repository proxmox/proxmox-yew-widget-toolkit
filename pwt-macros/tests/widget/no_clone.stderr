error[E0277]: the trait bound `Foo: Clone` is not satisfied
  --> tests/widget/no_clone.rs:47:1
   |
47 | #[widget(pwt=crate, comp=FooComp, @input, @element)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Clone` is not implemented for `Foo`
   |
   = note: this error originates in the attribute macro `widget` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider annotating `Foo` with `#[derive(Clone)]`
   |
47 + #[derive(Clone)]
48 | #[widget(pwt=crate, comp=FooComp, @input, @element)]
   |
